File: dotnet/DesignPatterns/BehavioralDesignPatterns/StateDesignPattern/StateDesignPattern.Example1/ShippedState.cs
namespace StateDesignPattern.Example1;

/// <summary>
/// Concrete State - represents an order that has been shipped.
/// </summary>
public class ShippedState : IOrderState
{
    public void ProcessOrder(Order order)
    {
        Console.WriteLine("Cannot process. Order has already been shipped.");
    }

    public void ShipOrder(Order order)
    {
        Console.WriteLine("Order has already been shipped.");
    }

    public void DeliverOrder(Order order)
    {
        Console.WriteLine("Order is being delivered...");
        order.SetState(new DeliveredState());
        Console.WriteLine($"Order state changed to: {order.GetCurrentStateName()}");
    }

    public void CancelOrder(Order order)
    {
        Console.WriteLine("Cannot cancel. Order has already been shipped.");
    }

    public string GetStateName() => "Shipped";
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/StateDesignPattern/StateDesignPattern.Example1/src/ShippedState.ts
import { IOrderState } from './IOrderState';
import { Order } from './Order';
import { DeliveredState } from './DeliveredState';

export class ShippedState implements IOrderState {
  processOrder(order: Order): void {
    console.log('Cannot process. Order has already been shipped.');
  }

  shipOrder(order: Order): void {
    console.log('Order has already been shipped.');
  }

  deliverOrder(order: Order): void {
    console.log('Order is being delivered...');
    order.setState(new DeliveredState());
    console.log(`Order state changed to: ${order.getCurrentStateName()}`);
  }

  cancelOrder(order: Order): void {
    console.log('Cannot cancel. Order has already been shipped.');
  }

  getStateName(): string {
    return 'Shipped';
  }
}
